# PISA 2012 Data Exploration
## by Anna Pedroni



From the [PISA Data Visualization Contest](http://www.oecd.org/pisa/pisaproducts/datavisualizationcontest.htm) webpage:
> PISA is a worldwide study developed by the Organisation for Economic Co-operation and Development (OECD) which examines the skills of 15-year-old school students around the world. The study assesses students’ mathematics, science, and reading skills and contains a wealth of information on students’ background, their school and the organisation of education systems. For most countries, the sample is around 5,000 students, but in some countries the number is even higher. In total, the PISA 2012 dataset contains data on 485 490 pupils.

A detailed description of the methodology of the PISA surveys can be found in the [PISA 2012 Technical Report](http://www.oecd.org/pisa/pisaproducts/PISA-2012-technical-report-final.pdf).


## A few points of interest.

### PISA:

- is an age-based survey, assessing 15-year-old students in school in grade 7 or higher. These students are approaching the end of compulsory schooling in most participating countries, and school enrolment at this level is close to universal in almost all OECD countries;
- take a literacy perspective, which focuses on the extent to which students can apply the knowledge and skills they have learned and practised at school when confronted with situations and challenges for which that knowledge may be relevant;
- allows for the assessment of additional cross-curricular competencies [...]. For 2012 a computer-delivered assessment of mathematics and problem solving was added, along with an assessment of financial literacy;
- uses Student Questionnaires to collect information from students on various aspects of their home, family and school background;
- uses School Questionnaires to collect information from schools about various aspects of organisation and educational provision in schools
- uses Parent Questionnaires administered to the parents of the students participating in PISA (in 11 countries for the 2012 survey).

## Focus and Partecipation

PISA 2012, the fifth PISA survey covered **reading, mathematics, science, problem solving and financial literacy** with a primary focus on mathematics.

It was conducted in 34 OECD countries and 31 partner countries/economies.
All 65 countries/economies completed the paper-based tests, with assessments lasting a total of two hours for each student.

An additional 40 minutes were devoted to the computer-based assessment of
- problem solving, in 44 countries/economies;
- mathematics and reading, in 32 countries/economies;
- financial literacy, in 18 countries/economies.

The full list of participants can be found [here](http://www.oecd.org/pisa/aboutpisa/pisa-2012-participants.htm).

Whether they took part in the additional computer-based assessments or not can be found in the [Technical Report](http://www.oecd.org/pisa/pisaproducts/PISA-2012-technical-report-final.pdf) at pp.23-24.

### Accordingly to the PISA Technical report, using the data from Student, Parent and School Questionnaires, analyses linking contextual information with student achievement could address:


- differences between countries in the relationships between student-level factors (such as gender and socio-economic background) and achievement;
- differences in the relationships between school-level factors and achievement across countries;
- differences in the proportion of variation in achievement between (rather than within) schools, and differences in this value across countries;
- differences between countries in the extent to which schools moderate or increase the effects of individual-level student factors and student achievement;
- differences in education systems and national context that are related to differences in student achievement across countries; and
- through links to PISA 2000, PISA 2003, PISA 2006 and PISA 2009, changes in any or all of these relationships over time.

## Preliminary Wrangling

The links to 2 files where provided with the [Udacity description of the databases](https://video.udacity-data.com/topher/2019/April/5ca78b26_dataset-project-communicate-data-findings/dataset-project-communicate-data-findings.pdf) for the project:

- PISA Data: pisa2012.csv(.zip) -file with the data about the 485 490 pupils
- PISA Data Dictionary: pisa.dict2012.csv - name of the columns and their description

### Download the data:

In [2]:
# import packages to download files and manage folders
import os
import requests
import zipfile

In [48]:
# create a folder and get the files

folder_name = 'PISA_data'

if not os.path.exists(folder_name):
    os.makedirs(folder_name)

urls = ['https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip',
        'https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisadict2012.csv']
         
for url in urls:
    response = requests.get(url)
    file_name = url.split('/')[-1]
    with open(os.path.join(folder_name, file_name), mode='wb') as file:
        file.write(response.content)

In [54]:
# unzip the PISA data 

file_name = 'pisa2012.csv.zip'

with zipfile.ZipFile(os.path.join(folder_name, file_name)) as data_zip:
    data_zip.extractall(folder_name)
print('Unzipped')

# remove the .zip file
os.remove(os.path.join(folder_name, file_name))
print('Removed')

Unzipped
Removed


## Import the data and have a look

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

### The dictionary of variables

In [2]:
# load the data and have a look
pisa_variables = pd.read_csv('PISA_data\pisadict2012.csv', encoding='latin-1', dtype='unicode')
pisa_variables.head()

Unnamed: 0                                                  x
0        CNT                           Country code 3-character
1   SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2    STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3       OECD                                       OECD country
4         NC                       National Centre 6-digit Code

In [3]:
pisa_variables.shape

(635, 2)

In [4]:
pisa_variables.columns = ['code', 'x']

In [5]:
pisa_variables['x'].values

array(['Country code 3-character',
       'Adjudicated sub-region code 7-digit code (3-digit country code + region ID + stratum ID)',
       'Stratum ID 7-character (cnt + region ID + original stratum ID)',
       'OECD country', 'National Centre 6-digit Code',
       'School ID 7-digit (region ID + stratum ID + 3-digit school ID)',
       'Student ID', 'International Grade', 'National Study Programme',
       'Birth - Month', 'Birth -Year', 'Gender', 'Attend <ISCED 0>',
       'Age at <ISCED 1>', 'Repeat - <ISCED 1>', 'Repeat - <ISCED 2>',
       'Repeat - <ISCED 3>', 'Truancy - Late for School',
       'Truancy - Skip whole school day',
       'Truancy - Skip classes within school day', 'At Home - Mother',
       'At Home - Father', 'At Home - Brothers', 'At Home - Sisters',
       'At Home - Grandparents', 'At Home - Others',
       'Mother<Highest Schooling>',
       'Mother Qualifications - <ISCED level 6>',
       'Mother Qualifications - <ISCED level 5A>',
       'Mother Qualifi

### The main file: data of the PISA 2012 survey

In [6]:
# load the PISA 2012 survey data (spoiler: it would take a lot! of time on my laptop)
pisa_data = pd.read_csv('PISA_data\pisa2012.csv', encoding='latin-1', dtype='unicode')
pisa_data.head()

Unnamed: 0      CNT SUBNATIO  STRATUM      OECD       NC SCHOOLID STIDSTD  \
0          1  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00001   
1          2  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00002   
2          3  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00003   
3          4  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00004   
4          5  Albania  0080000  ALB0006  Non-OECD  Albania  0000001   00005   

  ST01Q01 ST02Q01  ... W_FSTR75 W_FSTR76 W_FSTR77 W_FSTR78 W_FSTR79 W_FSTR80  \
0      10       1  ...  13.7954  13.9235  13.1249  13.1249   4.3389  13.0829   
1      10       1  ...  13.7954  13.9235  13.1249  13.1249   4.3389  13.0829   
2       9       1  ...  12.7307  12.7307  12.7307  12.7307   4.2436  12.7307   
3       9       1  ...  12.7307  12.7307  12.7307  12.7307   4.2436  12.7307   
4       9       1  ...  12.7307  12.7307  12.7307  12.7307   4.2436  12.7307   

  WVARSTRR VAR_UNIT SENWGT_STU  VER_STU  
0       19        1     0.2098  22NOV13  
1       19        1     0.2098  22NOV13  
2       19        1     0.1999  22NOV13  
3       19        1     0.1999  22NOV13  
4       19        1     0.1999  22NOV13  

[5 rows x 636 columns]

In [7]:
pisa_data.shape

(485490, 636)

### What is the structure of your dataset?

The main dataset contains 485490 rows, each one representing one student, and 636 columns with coded names.

A second files provide the dictionary for the criptic column names. There we see that each row of the main dataset
- starts with 6 columns about the location of the school attended by the student (e.g. country code, subregion, school ID)
- follows with the sudent ID column and a lot of personal information about the student and their family, their economic situation and experience in school (both emotional and with the school teaching), and a lot more.
- ends with their PLAUSIBLE results at the PISA tests (math and math subscales, science and reading; NO financial literacy) and the weight of the entry within their country, and
- the data on which the entry was created.

### What is/are the main feature(s) of interest in your dataset?

The description given above is a very summary one, and the dataset is truly fascinating for the possibilities of analysis it offers, even if it contains only the results for the main survey (the one used in all the countries).

Sadly, here I need to cut on the information I handle, because of scarcity of resources (time and computer memory). Anyway a detailed analysis is (luckily) not requested.

As said PISA 2012 focussed on mathematics, so in the dataset there are plausible scores about subset of mathematic competencies.

Moreover, Countries that participated in the survey have different languages, but also different cultures and writing systems.
I will select a subset of countries such as their writing system belongs to one of these groups:
  - alphabetic, with a highly phonemic orthography (or shallow orthography): Spanish, Finnish, Italian, German *
  - alphabetic, with a more complex relationship between orthography and pronunciation, a deeper orthography: English, French, Arabic *
  - logographic: Chinese, Japanese ([Korean](https://en.wikipedia.org/wiki/Korean_language#Writing_system), but keeping in mind that Korean logograms, still studied in school, as been long replaced in use by Korean alphabet) *


\* *I wanted to make the distinction between languages "that are written how they are pronunced", languages "that you cannot guess how a word is written just by hearing it" and languages that use ideograms... [Wikipedia](https://en.wikipedia.org/wiki/Writing_system) and the article [Getting to the bottom of orthographic depth](https://link.springer.com/article/10.3758/s13423-015-0835-2#Sec7) helped me putting things down a little more precisely.*


### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

I will keep for sure the mathematic and reading scores, as well as the language(s) spoken by the students and the language in which the test was administered (*TESTLANG* is the column heading).

Since the mathematic scores have subcategories, it would probably be interesting to have a look at them separately.

*(More about the scores and proficiency levels in the next cells)*

There is a wealth of information about the students' background as well, and it seems like a good idea to see if there are influential factors there, for instance the *index of economic, social and cultural status (column "ESCS" in the dataset)*; but to really understand what information I have that should be used, I first need to trim the database and explore it better.

### Proficiency levels

In the [Technical Report](http://www.oecd.org/pisa/pisaproducts/PISA-2012-technical-report-final.pdf) (from p.296 on) it is possible to find the scales used to record the proficiency for the different areas:

- Mathematics: >1, 1, 2, 3, 4, 5, 6 (highest level) *

- Reading and Science: I couldn't find the scales in the document. There is, however, indicated that the scales are just one for each discipline and are respectively the same as 2009 and 2006.
  - Reading: 1a, 1b, 2, 3, 4, 5, 6 (highest level)
  - Science: 1, 2, 3, 4, 5, 6 (highest level)

All this proficiency levels are provided with a description of the competencies associated with them.

(problem solving and financial literacy are also described, but the relative plausible scores are not in the dataset)

*NB. this is the main scale for mathematics, and also for all the subscales of different competencies tested (e.g.: "formulating situations mathematically", "employing mathematical concepts, facts, ...")

### However, the results are not given by those scales!

We only have PLAUSIBLE VALUES:
From the Technical report:
>Plausible values

>As with all item response scaling models, student proficiencies (or measures) are not observed; they are missing data
that must be inferred from the observed item responses. There are several possible alternative approaches for making
this inference. PISA uses the imputation methodology usually referred to as plausible values (PVs). PVs are a selection of
likely proficiencies for students that attained each score. For each scale and subscale, five plausible values per student
are included in the international database.
Using item parameters anchored at their estimated values from the international calibration, the plausible values are
random draws from the marginal posterior of the latent distribution for each student. 

### About the plausible scales (p.156):

> Sixty-five plausible values, five for each of the 13 PISA 2012 scales are included in the PISA 2012 database. PV1MATH
to PV5MATH are for mathematical literacy; PV1SCIE to PV5SCIE for scientific literacy, PV1READ to PV5READ for
reading literacy, PV1CPRO to PV5CPRO for computer problem solving assessment, PV1CMAT to PV5CMAT for the
computer-based mathematics assessment and PV1CREA to PV5CREA for digital reading assessment. For the four
mathematics content subscales, change and relationships, quantity, space and shape, uncertainty and data, the plausible
values variables are PV1MACC to PV5MACC, PV1MACQ to PV5MACQ, PV1MACS to PV5MACS, and PV1MACU to
PV5MACU respectively. For the three mathematics process subscales employ, formulate and interpret, the plausible
values variables are PV1MAPE to PV5MAPE, PV1MAPF to PV5MAPF, and PV1MAPI to PV5MAPI respectively

### My understanding about these evil PLAUSIBLE SCORES

Reading through the Technical report, I've come up with this layman explanation:

For every student, for every scale and subscale, the results of the tests have been taken and used to compute 5 plausible results that student could have reached if they had taken all the PISA tests, not only the subset contained in the booklet they tackled.
This has been done for accuracy in later estimating the parameters of the population (all the students of a Country). Accuracy of those parameters is also the reason it would be better to use the weight of each student and also to repeat the calculations 5 times per scale (one for each plausible score column).

The plausible score is given in a scale that can be converted to the relative proficiency level according to the bands provided in the Technical report:

##### Main mathematical literacy levels and subscales as well - p. 297 of the Technical report
Level : Score points on the PISA scale
- 6 : Above 669.3
- 5 : From 607.0 to less than 669.3
- 4 : From 544.7 to less than 607.0
- 3 : From 482.4 to less than 544.7
- 2 : From 420.1 to less than 482.4
- 1 : From 357.8 to less than 420.1
- Below level 1 : Below 357.8


##### Reading literacy performance band definitions on the PISA scale - p. 265 of the [PISA 2009 Technical report](https://www.oecd-ilibrary.org/docserver/9789264167872-en.pdf?expires=1587756729&id=id&accname=guest&checksum=3C4CAAFFDA33331CACE78A9B3A852FA5)
Level : Score points on the PISA scale
- 6 : Higher than 698.32
- 5 : Higher than 625.61 and less than or equal to 698.32
- 4 : Higher than 552.89 and less than or equal to 625.61
- 3 : Higher than 480.18 and less than or equal to 552.89
- 2 : Higher than 407.47 and less than or equal to 480.18
- 1a : Higher than 334.75 and less than or equal to 407.47
- 1b : 262.04 to less than or equal to 334.75


##### Scientific literacy performance band definitions on the PISA scale - p. 293 of the [PISA 2006 Technical report](https://www.oecd.org/pisa/data/42025182.pdf)
Level : Score points on the PISA scale
- 6 : Above 707.9
- 5 : 633.3 to 707.9
- 4 : 558.7 to 633.3
- 3 : 484.1 to 558.7
- 2 : 409.5 to 484.1
- 1 : 334.9 to 409.5

### About the weights

Every row/student has a weight (more than one, to be precise, but..).

The reason for it resides, if I got it right, in the sampling process of subregions, schools and students done in every Country and the reason to use it (them) is, in the end, to better represent the country parameters.

I'm planning to regroup the students by different parameters (primary first language = language at school = language of the PISA test), so they won't be divided by Country and won't rapresent country level proficiency.

Therefore I think it would be wrong to use the weights within this project.

## More Wrangling

As said, I will keep only the data about the countries speaking one or more of these lenguages: Spanish, Finnish, Italian, German, English, French, Arabic, Chinese, and Japanese.

The countries in the database are:

In [8]:
# find the list of countries contained
pisa_data.CNT.unique()

array(['Albania', 'United Arab Emirates', 'Argentina', 'Australia',
       'Austria', 'Belgium', 'Bulgaria', 'Brazil', 'Canada',
       'Switzerland', 'Chile', 'Colombia', 'Costa Rica', 'Czech Republic',
       'Germany', 'Denmark', 'Spain', 'Estonia', 'Finland', 'France',
       'United Kingdom', 'Greece', 'Hong Kong-China', 'Croatia',
       'Hungary', 'Indonesia', 'Ireland', 'Iceland', 'Israel', 'Italy',
       'Jordan', 'Japan', 'Kazakhstan', 'Korea', 'Liechtenstein',
       'Lithuania', 'Luxembourg', 'Latvia', 'Macao-China', 'Mexico',
       'Montenegro', 'Malaysia', 'Netherlands', 'Norway', 'New Zealand',
       'Peru', 'Poland', 'Portugal', 'Qatar', 'China-Shanghai',
       'Perm(Russian Federation)', 'Florida (USA)', 'Connecticut (USA)',
       'Massachusetts (USA)', 'Romania', 'Russian Federation',
       'Singapore', 'Serbia', 'Slovak Republic', 'Slovenia', 'Sweden',
       'Chinese Taipei', 'Thailand', 'Tunisia', 'Turkey', 'Uruguay',
       'United States of America', 'Vietn

By the way, these are the full names, not the 3 character Country codes as indicated by the data dictionary.. which is good, because it is easier for me to check on the Wikipedia pages the spoken languages, for the countries I have doubts about.

I'm a bit puzzled by the presence of United States of America and, separately, Florida, Connecticut and Massachusetts. Assuming that all the data included in the dataset are valid (the survey data have been revised several times, if we believe the Technical report -and we have to) the reason is possibly to be find in the sampling choices and volunteering of schools to conduct the survey. Provided there are no duplicates, it is no worries to me.

I will keep these 37 Countries/economies:
'United Arab Emirates', 'Argentina', 'Australia', 'Austria', 'Belgium', 'Canada', 'Switzerland', 'Chile', 'Colombia', 'Costa Rica', 'Germany', 'Spain', 'Finland', 'France', 'United Kingdom', 'Hong Kong-China', 'Ireland', 'Italy', 'Japan', 'Korea', 'Liechtenstein', 'Luxembourg', 'Macao-China', 'Mexico', 'New Zealand', 'Peru', 'Qatar', 'China-Shanghai', 'Florida (USA)', 'Connecticut (USA)', 'Massachusetts (USA)', 'Singapore', 'Chinese Taipei', 'Tunisia', 'Uruguay', 'United States of America', 'Vietnam'

In [9]:
# list of Countries to keep
to_keep = ['United Arab Emirates', 'Argentina', 'Australia', 'Austria', 'Belgium', 'Canada', 'Switzerland', 'Chile', 'Colombia', 'Costa Rica', 'Germany', 'Spain', 'Finland', 'France', 'United Kingdom', 'Hong Kong-China', 'Ireland', 'Italy', 'Japan', 'Korea', 'Liechtenstein', 'Luxembourg', 'Macao-China', 'Mexico', 'New Zealand', 'Peru', 'Qatar', 'China-Shanghai', 'Florida (USA)', 'Connecticut (USA)', 'Massachusetts (USA)', 'Singapore', 'Chinese Taipei', 'Tunisia', 'Uruguay', 'United States of America', 'Vietnam']


# select only the desired Countries:
# reusing the df name in order not to use up too much memory
pisa_data = pisa_data.loc[pisa_data.CNT.isin(to_keep)]

# clean memory from trial results (answer "y" to prompt)
%reset Out

Once deleted, variables cannot be recovered. Proceed (y/[n])? y
Flushing output cache (6 entries)


In [10]:
pisa_data.shape

(314831, 636)

We had 485490 rows, now they are **314831** (we dismissed roughly 35% of the rows).

Time to eliminate some of the columns. I'm not ready to dismiss most of the columns, even if at the end I will use only a few of the 636 that are in the dataset.

I'm quite confident in dismissing the following two chunks though (I'm not bothering with picking up single columns):

 - from : *EC04Q01A	Acquired skills - Find job info - Yes, at school*
 
 to: *EC04Q06C	Acquired skills - Student financing - No, never*

 
 - from: *W_FSTR1	FINAL STUDENT REPLICATE BRR-FAY WEIGHT1*
 
 to: *VAR_UNIT	RANDOMLY ASSIGNED VARIANCE UNIT*


In [11]:
# delete the columns listed above
pisa_data.drop(pisa_data.loc[:, 'EC04Q01A':'EC04Q06C'].columns, axis=1, inplace=True)
pisa_data.drop(pisa_data.loc[:, 'W_FSTR1':'VAR_UNIT'].columns, axis=1, inplace=True)
# since I'm doing this, I'll drop also the first, unnamed, column
pisa_data.drop(pisa_data.iloc[:,0:1].columns, axis=1, inplace=True)

In [12]:
# looking a bit better?
pisa_data.shape

(314831, 535)

From 636 columns to 535: probably still 510-520 useless columns, but 101 column better.

I'll save my subset of the database as it is now, so that if needed, I can restart running the cells from the last one before the next section, "Select variables".

In [14]:
pisa_data.dtypes

CNT           object
SUBNATIO      object
STRATUM       object
OECD          object
NC            object
               ...  
PV4SCIE       object
PV5SCIE       object
W_FSTUWT      object
SENWGT_STU    object
VER_STU       object
Length: 535, dtype: object

In [15]:
# save the dataset, keep ',' as separator, keep 'utf8' as encoding
pisa_data.to_csv('PISA_data\selected_countries_subset.csv', index=False, encoding='latin-1')
print('Saved')

Saved


In [22]:
# clean a bit of memory
pisa_data = ''
%reset Out

Once deleted, variables cannot be recovered. Proceed (y/[n])? y
Flushing output cache (9 entries)


In [20]:
# in the case I need to restart the kernel or the computer
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

%matplotlib inline

# reload the dictionary
pisa_variables = pd.read_csv('PISA_data\pisadict2012.csv', encoding='latin-1', dtype='unicode')
pisa_variables.columns = ['code', 'x']


# load my selected Countries data
pisa_subset = pd.read_csv('PISA_data\selected_countries_subset.csv', encoding='latin-1', dtype='unicode')
#have a look
pisa_subset.head()

CNT SUBNATIO  STRATUM      OECD                     NC  \
0  United Arab Emirates  7840000  ARE0762  Non-OECD  United Arab Emirates    
1  United Arab Emirates  7840000  ARE0762  Non-OECD  United Arab Emirates    
2  United Arab Emirates  7840000  ARE0762  Non-OECD  United Arab Emirates    
3  United Arab Emirates  7840000  ARE0762  Non-OECD  United Arab Emirates    
4  United Arab Emirates  7840000  ARE0762  Non-OECD  United Arab Emirates    

  SCHOOLID STIDSTD ST01Q01 ST02Q01 ST03Q01  ...   PV4READ   PV5READ   PV1SCIE  \
0  0000001   00001      10       1      08  ...  310.3345  241.3672  397.2771   
1  0000001   00002      10       1      06  ...  198.2226  243.1315  301.5106   
2  0000001   00003      10       1      09  ...  251.0707  322.4439   293.771   
3  0000001   00004       9       1      12  ...  214.1813  252.6746  329.2055   
4  0000001   00005      10       1      03  ...  301.6735   249.547  378.1611   

    PV2SCIE   PV3SCIE   PV4SCIE   PV5SCIE W_FSTUWT SENWGT_STU  VER_STU  
0  387.9523  375.8299  365.5725  351.5852   2.8838      0.071  22NOV13  
1  267.0086  315.4979  289.3883  305.2406   2.8838      0.071  22NOV13  
2  278.8512  200.5222  234.0918  327.3405   2.8838      0.071  22NOV13  
3  298.4334  281.6486  319.8806  318.9482   2.8838      0.071  22NOV13  
4  405.2033   389.351  408.0007  410.7982   2.8838      0.071  22NOV13  

[5 rows x 535 columns]

### Select variables

In [21]:
pisa_subset.shape

(314831, 535)

In [48]:
pisa_subset.CNT.count()

314831

pisa_variables that look interesting:

'Country code 3-character',
'OECD country',
'Student ID',
'Gender',
'Mother<Highest Schooling>',
'Mother Current Job Status',
'Father Qualifications - <ISCED level 6>',
'Father Current Job Status',
'Country of Birth International - Self',
'First language learned',
'Age started learning <test language>',
'Language spoken - Mother',
'Language spoken - Father',
'Language spoken - Schoolmates',
'Acculturation - Enjoy <Host Culture> Friends',
'Acculturation - Enjoy <Heritage Culture> Friends',
'Acculturation - Enjoy <Host Culture> Celebrations',
'Acculturation - Enjoy <Heritage Culture> Celebrations',
'Acculturation - Spend Time with <Host Culture> Friends',
'Acculturation - Spend Time with <Heritage Culture> Friends',
'Acculturation - Participate in <Host Culture> Celebrations',
'Acculturation - Participate in <Heritage Culture> Celebrations',
'Standard or simplified set of booklets',
'Attitude towards School: Learning Outcomes',
'Attitude towards School: Learning Activities',
'Sense of Belonging to School',
"Mathematics Teacher's Classroom Management",
'Cognitive Activation in Mathematics Lessons',
'Instrumental Motivation for Mathematics',
'Mathematics Interest',
'Mathematics Work Ethic',
'Mathematics Teacher’s Support'
'Mathematics Self-Concept',
'Teacher-Student Relations',
'Subjective Norms in Mathematics',
'Index of economic, social and cultural status',
'Home educational resources',
'Highest educational level of parents',
'Preference for Heritage Language in Conversations with Family and Friends',
'Language at home (3-digit code)',
'Preference for Heritage Language in Language Reception and Production',
'Plausible value 1 in mathematics',
'Plausible value 1 in content subscale of math - Change and Relationships',
'Plausible value 1 in content subscale of math - Quantity',
'Plausible value 1 in content subscale of math - Space and Shape',
'Plausible value 1 in content subscale of math - Uncertainty and Data',
'Plausible value 1 in process subscale of math - Employ',
'Plausible value 1 in process subscale of math - Formulate',
'Plausible value 1 in process subscale of math - Interpret',
'Plausible value 1 in reading',
'Plausible value 1 in science'

#### There would be more, but already like this they are more than I will use


In [73]:
descriptions = ['Country code 3-character',
                'OECD country',
                'Student ID',
                'Gender',
                'Mother<Highest Schooling>',
                'Mother Current Job Status',
                'Father Qualifications - <ISCED level 6>',
                'Father Current Job Status',
                'Country of Birth International - Self',
                'First language learned',
                'Age started learning <test language>',
                'Language spoken - Mother',
                'Language spoken - Father',
                'Language spoken - Schoolmates',
                'Acculturation - Enjoy <Host Culture> Friends',
                'Acculturation - Enjoy <Heritage Culture> Friends',
                'Acculturation - Enjoy <Host Culture> Celebrations',
                'Acculturation - Enjoy <Heritage Culture> Celebrations',
                'Acculturation - Spend Time with <Host Culture> Friends',
                'Acculturation - Spend Time with <Heritage Culture> Friends',
                'Acculturation - Participate in <Host Culture> Celebrations',
                'Acculturation - Participate in <Heritage Culture> Celebrations',
                'Standard or simplified set of booklets',
                'Attitude towards School: Learning Outcomes',
                'Attitude towards School: Learning Activities',
                'Sense of Belonging to School',
                "Mathematics Teacher's Classroom Management",
                'Cognitive Activation in Mathematics Lessons',
                'Instrumental Motivation for Mathematics',
                'Mathematics Interest',
                'Mathematics Work Ethic',
                "Mathematics Teacher's Support",
                'Mathematics Self-Concept',
                'Teacher Student Relations',
                'Subjective Norms in Mathematics',
                'Index of economic, social and cultural status',
                'Home educational resources',
                'Highest educational level of parents',
                'Preference for Heritage Language in Conversations with Family and Friends',
                'Language at home (3-digit code)',
                'Preference for Heritage Language in Language Reception and Production',
                'Plausible value 1 in mathematics',
                'Plausible value 1 in content subscale of math - Change and Relationships',
                'Plausible value 1 in content subscale of math - Quantity',
                'Plausible value 1 in content subscale of math - Space and Shape',
                'Plausible value 1 in content subscale of math - Uncertainty and Data',
                'Plausible value 1 in process subscale of math - Employ',
                'Plausible value 1 in process subscale of math - Formulate',
                'Plausible value 1 in process subscale of math - Interpret',
                'Plausible value 1 in reading',
                'Plausible value 1 in science']

In [74]:
#get the column names
columns_to_keep = pisa_variables[pisa_variables['x'].isin(descriptions)]['code'].tolist()

In [78]:
pisa_chosen_var = pisa_subset[columns_to_keep]

In [80]:
pisa_chosen_var.columns = descriptions

In [81]:
pisa_chosen_var

Country code 3-character OECD country Student ID  Gender  \
0          United Arab Emirates     Non-OECD      00001    Male   
1          United Arab Emirates     Non-OECD      00002    Male   
2          United Arab Emirates     Non-OECD      00003    Male   
3          United Arab Emirates     Non-OECD      00004    Male   
4          United Arab Emirates     Non-OECD      00005    Male   
...                         ...          ...        ...     ...   
314826                  Vietnam     Non-OECD      04955  Female   
314827                  Vietnam     Non-OECD      04956    Male   
314828                  Vietnam     Non-OECD      04957    Male   
314829                  Vietnam     Non-OECD      04958    Male   
314830                  Vietnam     Non-OECD      04959  Female   

                    Mother<Highest Schooling>  \
0                           <ISCED level 3A>    
1                           <ISCED level 3A>    
2                                         NaN   
3                           <ISCED level 3A>    
4                           <ISCED level 3A>    
...                                       ...   
314826                       <ISCED level 2>    
314827  She did not complete <ISCED level 1>    
314828                       <ISCED level 1>    
314829                       <ISCED level 2>    
314830  She did not complete <ISCED level 1>    

                 Mother Current Job Status  \
0       Other (e.g. home duties, retired)    
1       Other (e.g. home duties, retired)    
2       Other (e.g. home duties, retired)    
3       Other (e.g. home duties, retired)    
4       Other (e.g. home duties, retired)    
...                                    ...   
314826  Other (e.g. home duties, retired)    
314827  Other (e.g. home duties, retired)    
314828        Working full-time <for pay>    
314829  Other (e.g. home duties, retired)    
314830  Other (e.g. home duties, retired)    

       Father Qualifications - <ISCED level 6>  \
0                                          Yes   
1                                          Yes   
2                                           No   
3                                          NaN   
4                                          NaN   
...                                        ...   
314826                                      No   
314827                                      No   
314828                                      No   
314829                                      No   
314830                                      No   

                 Father Current Job Status  \
0             Working full-time <for pay>    
1       Not working, but looking for a job   
2              Working part-time <for pay>   
3             Working full-time <for pay>    
4       Other (e.g. home duties, retired)    
...                                    ...   
314826  Other (e.g. home duties, retired)    
314827  Other (e.g. home duties, retired)    
314828        Working full-time <for pay>    
314829  Other (e.g. home duties, retired)    
314830        Working full-time <for pay>    

       Country of Birth International - Self First language learned  ...  \
0                              Other country                    NaN  ...   
1                            Country of test                    NaN  ...   
2                            Country of test                    NaN  ...   
3                              Other country                    NaN  ...   
4                            Country of test                    NaN  ...   
...                                      ...                    ...  ...   
314826                       Country of test                    NaN  ...   
314827                       Country of test                    NaN  ...   
314828                       Country of test                    NaN  ...   
314829                       Country of test                    NaN  ...   
314830                       Country of test                    NaN  .

## Univariate Exploration

> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.

> Make sure that, after every plot or related series of plots, that you
include a Markdown cell with comments about what you observed, and what
you plan on investigating next.

### Discuss the distribution(s) of your variable(s) of interest. Were there any unusual points? Did you need to perform any transformations?

> Your answer here!

### Of the features you investigated, were there any unusual distributions? Did you perform any operations on the data to tidy, adjust, or change the form of the data? If so, why did you do this?

> Your answer here!

## Bivariate Exploration

> In this section, investigate relationships between pairs of variables in your
data. Make sure the variables that you cover here have been introduced in some
fashion in the previous section (univariate exploration).

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> Your answer here!

### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> Your answer here!

## Multivariate Exploration

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

### Talk about some of the relationships you observed in this part of the investigation. Were there features that strengthened each other in terms of looking at your feature(s) of interest?

> Your answer here!

### Were there any interesting or surprising interactions between features?

> Your answer here!

> At the end of your report, make sure that you export the notebook as an
html file from the `File > Download as... > HTML` menu. Make sure you keep
track of where the exported file goes, so you can put it in the same folder
as this notebook for project submission. Also, make sure you remove all of
the quote-formatted guide notes like this one before you finish your report!